hazelcast/src/main/java/com/hazelcast/jet/impl/MasterJobContext.java

```diff
@@ -596,10 +596,17 @@ private void onStartExecutionComplete(Throwable error, Collection<Entry<MemberIn
                     .whenCompleteAsync(withTryCatch(logger, (r, e) -> finalizeJob(finalError)));
         } else {
             if (error instanceof ExecutionNotFoundException) {
-                // If the StartExecutionOperation didn't find the execution, it means that we must have cancelled it.
-                // Let's pretend that the StartExecutionOperation returned JobTerminateRequestedException
-                assert requestedTerminationMode != null && !requestedTerminationMode.isWithTerminalSnapshot();
-                error = new JobTerminateRequestedException(requestedTerminationMode);
+                // If the StartExecutionOperation didn't find the execution, it means that it was cancelled.
+                if (requestedTerminationMode != null) {
+                    // This cancellation can be because the master cancelled it upon to a user
+                    // request. Let's pretend that case as the StartExecutionOperation returned
+                    // JobTerminateRequestedException.
```

Comment:
```suggestion
                    // This cancellation can be because the master cancelled it. If that's the case, convert the exception
                    // to JobTerminateRequestedException.
```